crates/matrix-sdk/src/client/mod.rs

```diff
@@ -710,18 +714,20 @@ impl Client {
     #[instrument(skip(self, user, password))]
     pub async fn login(
         &self,
-        user: impl AsRef<str>,
-        password: &str,
+        user: UserIdentifier<'_>,
+        password: impl AsRef<str>,
         device_id: Option<&str>,
         initial_device_display_name: Option<&str>,
     ) -> Result<login::v3::Response> {
         let homeserver = self.homeserver().await;
-        info!(homeserver = homeserver.as_str(), user = user.as_ref(), "Logging in");
+        info!(
+            homeserver = homeserver.as_str(),
+            user = format!("{:?}", user).as_str(),
+            "Logging in"
+        );
 
-        let login_info = login::v3::LoginInfo::Password(login::v3::Password::new(
-            UserIdentifier::UserIdOrLocalpart(user.as_ref()),
-            password,
-        ));
+        let login_info =
+            login::v3::LoginInfo::Password(login::v3::Password::new(user, password.as_ref()));
```

Comment:
for reviewers: this is the real core of the change. Everything else is basically adapting to this, aside from some minor details.